packages/firestore/src/core/sync_engine.ts

```diff
@@ -373,7 +377,26 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
    */
   async write(batch: Mutation[], userCallback: Deferred<void>): Promise<void> {
     this.assertSubscribed('write()');
-    const result = await this.localStore.localWrite(batch);
+
+    let result: LocalWriteResult;
+    try {
+      result = await this.localStore.localWrite(batch);
+    } catch (e) {
+      // If we can't persist the mutation, we reject the user callback and don't
+      // send the mutation. The user can then retry the write. We currently do
+      // not use `enqueueRetryable()` for writes since this would require us to
+      // move all view computation logic to `enqueueRetryable()`. Otherwise,
+      // this write should be surfaced in all subsequent operations.
+      logError(LOG_TAG, 'Dropping write that cannot be persisted: ' + e);
+      userCallback.reject(
+        new FirestoreError(
+          Code.UNKNOWN,
+          'Query allocation failed with environment error: ' + e
```

Comment:
"Query allocation" seems like the wrong description of this error. Also, since this is catching everything, this isn't necessarily something to do with the environment. "Failed to enqueue a write in the local database"?

The catch here is also overbroad: this will catch internal assertion errors and the like as well. Taking an approach like the one I suggested in #2919 would make it possible to catch just errors related to IndexedDB.

Also, once we know the error is specifically about IndexedDB, I think the code can probably become `DATA_LOSS`.

Reply:
Sorry, "Query allocation" was the error message for my next PR.

I updated the PR with the changes in #2919 and only throw this error for IndexedDbTransactionErrors.  I am a little uncertain whether we should just re-throw the error here or wrap it. For now, I decided to wrap it in a FirestoreError, as this makes it easier for our users to share error handling logic between the various backend errors and the new IndexedDB error.